megatron/checkpointing.py

```diff
@@ -152,7 +152,22 @@ def save_ds_checkpoint(iteration, model, neox_args):
         logits = do_forward_pass(neox_args=neox_args, model=model)
         sd['checkpoint_validation_logits'] = logits
     
-    model.save_checkpoint(neox_args.save, client_state=sd)
+    # Derive tag 
+    # The tag is usually automatically derived in the same way in deepspeed
+    # As the save_checkpoint function does not return such derived tag, we 
+    # do it manually in order to use the subdirectory to save config files
+    tag = f"global_step{iteration}"
+
+    # save checkpoint
+    model.save_checkpoint(neox_args.save, tag=tag, client_state=sd)
+
+    # save config files
+    if  torch.distributed.get_rank() == 0 and neox_args.config_files is not None:
```

Comment:
should this be local_rank == 0? What if there isn't a shared filesystem

Reply:
Following reasoning. Hence I would like to keep it only for rank 0.

- also without shared file system the checkpoint is only complete with data from all ranks. There is no added benefit, you would not use only part of the checkpoint.
- duplicating data across the ranks then would require overwrites when moving the files to one location
- I have had at least one shared file system with problems writing to the same file from multiple ranks. This would avoid such multiple write handles. Deepspeed does it as well for latest file and the zero conversion script. Maybe should not be a problem, but it seems to be one some times.

Reply:
Deepspeed DOES write to one file at the same time from multiple ranks. To avoid confusion…